todo-list-backend/src/index.ts

```diff
@@ -0,0 +1,124 @@
+
+import {User} from "./types";
+import {InMemoryStorage, IStorage} from "./data"
+import {buildPermissions} from "./permissions";
+
+const typeDefs = gql`
+    type User {
+        name: String
+    }
+
+    type ToDo {
+        id: Int
+        name: String
+        done: Boolean
+        assignee: User
+    }
+
+    type Query {
+        readTodos: [ToDo]!
+    }
+
+    type Mutation {
+        login(name: String!, password: String!): String
+        createTodo(name: String!): ToDo
+        updateTodoName(id: Int!, name: String!): ToDo
+        updateTodoDone(id: Int!, done: Boolean!): ToDo
+        deleteTodo(id: Int!): ToDo
+        createUser(name: String! password: String!): User
+    }
+`;
+
+function buildResolvers(storage: IStorage, authSecret: string): any {
+    return {
+        Query: {
+            readTodos: (parent: any, args: any, context: any, info: any) => {
+                let user: User = context.user;
+                return storage.readTodos(user);
+            },
+        },
+
+        Mutation: {
+            login: (parent: any, args: any, context: any, info: any) => {
+                let user = storage.readUser(args.name);
+
+                if (!user || !user.checkPassword(args.password)) {
+                    return null;
+                }
+
+                let payload = {
+                    username: user.name
+                };
+
+                return sign(payload, authSecret, {expiresIn: "1 day"});
+            },
+            createTodo: (parent: any, args: any, context: any, info: any) => {
+                let user: User = context.user;
+                return storage.createTodo(user, args.name);
+            },
+            updateTodoName: (parent: any, args: any, context: any, info: any) => {
+                return storage.updateTodoName(context.user, args.id, args.name);
+            },
+            updateTodoDone: (parent: any, args: any, context: any, info: any) => {
+                return storage.updateTodoDone(context.user, args.id, args.done);
+            },
+            deleteTodo: (parent: any, args: any, context: any, info: any) => {
+                return storage.deleteTodo(context.user, args.id);
+            },
+            createUser: (parent: any, args: any, context: any, info: any) => {
+                return storage.createUser(args.name, args.password);
+            }
+        }
+    }
+}
+
+export function buildApolloServer(storage: IStorage, authSecret: string,
+                                  debugMode: boolean = false, context?: any): ApolloServer {
+
+    const resolvers = buildResolvers(storage, authSecret);
+    const schema = makeExecutableSchema({typeDefs, resolvers});
+    const permissions = buildPermissions(schema, authSecret, debugMode);
+    const finalizedSchema = applyMiddleware(schema, permissions);
+
+    return new ApolloServer({
+        schema: finalizedSchema,
+        context: ({req}) => {
+            if (context) {
+                return context;
```

Comment:
Interesting, so you want to overwrite context with a param to `buildApolloServer` here? The `context` of `apollo-server` is a function. For consistency I would check if the given `context` is a function and then call it.